_posts/2020-08-18-scaffolding-static-site-with-cdk.md

```diff
@@ -0,0 +1,243 @@
+---
+layout: post
+title: Scaffolding your AWS static site insfrastructure with CDK
+date: 2020-08-24
+tags: aws cdk static-site react angular cloud-development-kit infrastructure-as-code
+author: Tim Veletta
+---
+
+Many of client jobs we do at Mechanical Rock involve the creation of static sites so it often helps to have a simple, repeatable process for quickly scaffolding one. The AWS Cloud Development Kit (CDK) is a relatively new framework for defining your cloud infrastructure as code to be provisioned with CloudFormation. CDK allows developers of static sites to use the tools they are already familiar with, in this case TypeScript (or JavaScript), to define a template for their infrastructure that can be reused from one project to the next.
+
+In this post I will detail the process from creating an S3 bucket to host our sites assets through to provisioning certificates to enable HTTPS on our domain and everything in between using CDK. I do assume a some amount of familiarity with CDK, that you have gone through and completed the CDK setup as detailed in [the AWS Documentation](https://docs.aws.amazon.com/cdk/latest/guide/getting_started.html) and that you have a static site ready to deploy along with a custom domain name registered through AWS Route 53.
+
+The infrastructure we are deploying and the communication points are shown in the diagram below. We are able to break things into 3 distinct chunks that build on our infrastructure starting with an S3 static site. We add to it by implementing CloudFront for serving the site using a Content Delivery Network (CDN) and finally add our own secured domain.
+
+![CDK Hosting Infrastructure](/img/blog/cdk-static-site/cdk-hosting-infrastructure.png)
+
+## Dependencies
+
+For this Stack, we are going to need the following dependencies.
+
+```
+npm install @aws-cdk/aws-certificatemanager @aws-cdk/aws-cloudfront @aws-cdk/aws-cloudfront-origins @aws-cdk/aws-route53 @aws-cdk/aws-route53-patterns @aws-cdk/aws-route53-targets @aws-cdk/aws-s3
+```
+
+Just one thing to note, CDK versions change very frequently; this was all written and tested with CDK version **1.59.0**.
+
+## The Basic Site
+
+The first stage involves creating an S3 bucket and configuring it to serve the files using a S3 static site. We create a new function within our Stack to do this.
+
+```typescript
+private getS3Bucket(bucketName: string) {
+    const bucketConfig: BucketProps = {
+        bucketName,
+        websiteIndexDocument: "index.html",
+        websiteErrorDocument: "index.html",
+        publicReadAccess: true,
```

Comment:
If you're doing a SPA, you don't need do this since you're granting read to your OAI on line 79 of this post.

If you don't specify this, all public access is off by default, so you don't need to bother with L68-71

Reply:
I will add though, if you're doing something like [Hugo](https://gohugo.io) though, each page is rendered by a path, so you _do_ need a public read.

Reply:
I think its needed if you're serving the static site through S3 which is the initial stage but as soon as you add in CloudFront its no longer needed?

Reply:
Ah I see what you mean. It's done in stages rather than in one go.